공통 응답, 공통 예외 처리 설정

eggnee opened the pull request:
### ✅ What is this PR

- 공통 응답, 공통 예외 처리 설정


### 📄 Changes Made

- 공통 응답, 공통 예외 처리 설정


### 🙋🏻‍ Review Point

- 사용 방법 `TestController`, `SampleErrorCode`, `SampleException` 참고하시면 됩니다.
- 도메인마다 `~ErrorCode`, `~Exception` 만들어서 사용하시면 됩니다.

### 
- Issue #3

eggnee committed:
feat(#3): 공통 응답, 공통 예외 처리 설정

chs98412 commented on src/main/java/com/switcheese/server/common/response/ApiResponse.java:
```diff
@@ -0,0 +1,40 @@
+    return new ApiResponse<>(statusName, message, data);
+  }
+
+  public static <T> ApiResponse<T> created(T data) {
+    return new ApiResponse<>("CREATED", null, data);
+  }
+
+  public static <T> ApiResponse<T> ok(T data) {
+    return new ApiResponse<>("OK", null, data);
+  }
+
+  public static ApiResponse<Void> ok() {
+    return new ApiResponse<>("OK", null, null);
+  }
```
정상 상태일때 모든 http 코드가 200으로 갈 것 같은데 맞나용

eggnee replied:
네 맞아요 에러 났을 때도 http status 코드 자체는 200으로 갑니당

ghyen commented on src/main/java/com/switcheese/server/common/exception/BusinessExceptionHandler.java:
```diff
@@ -0,0 +1,25 @@
+  @ExceptionHandler(BusinessException.class)
+  public ApiResponse<Void> handleBusinessException(BusinessException ex) {
+    log.error(ex.getStatus().name(), ex.getMessage());
+    return ApiResponse.error(ex.getStatus().name(), ex.getMessage());
```
혹시 ex.getCode():에러코드는 안넣는 이유가 있나요?

eggnee replied:
ApiResponse error 메서드 자체에 code를 받고 있지 않아서요! 추가하는 게 좋을까요?

ghyen replied:
code도 포함시켜야지 프론트에서 처리하기 편할 것 같습니다!

ghyen reviewed (approved):
리뷰가 늦었습니다..ㅠ SampleCode 만들어주셔서 고마워요~!

eggnee committed:
feat(#3): 공통 응답 객체 제거 및 전역 예외 처리 구현

- ApiResponse와 같은 공통 응답 객체 제거 (중복성 문제 해결)
- 전역 예외 처리 시 ResponseEntity의 body에 ErrorResponse 반환
- 세분화된 전역 예외 처리 핸들러 구현

chs98412 commented on src/main/java/com/switcheese/server/common/exception/BusinessExceptionHandler.java:
```diff
@@ -1,25 +1,80 @@
-package com.switcheese.server.common.exception;
+package com.couponify.couponapi.exception;
 
```
zzzzz 수정해주세여

eggnee replied:
어라ㅋ

chs98412 commented on src/main/java/com/switcheese/server/common/exception/BusinessExceptionHandler.java:
```diff
@@ -1,25 +1,80 @@
+    return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
+        .body(errorResponse);
+  }
+
+  @ExceptionHandler(BusinessException.class)
+  public ResponseEntity<ErrorResponse> businessException(
+      BusinessException exception,
+      HttpServletRequest request
+  ) {
+    ErrorResponse errorResponse = ErrorResponse.fromBusinessException(request, exception);
+
+    log.error(errorResponse.toString());
+
+    return ResponseEntity.status(exception.getStatus())
+        .body(errorResponse);
```
동작은 하는거 확인했는데 Exception보다 밑에 있어도 어떻게 동작하는거져?

eggnee replied:
넹 순서는 상관 없어용

chs98412 commented on src/main/java/com/switcheese/server/common/exception/BusinessExceptionHandler.java:
```diff
@@ -1,25 +1,80 @@
-  public ApiResponse<Void> handleBusinessException(BusinessException ex) {
-    log.error(ex.getStatus().name(), ex.getMessage());
-    return ApiResponse.error(ex.getStatus().name(), ex.getMessage());
+  @ExceptionHandler({
+      HttpRequestMethodNotSupportedException.class,
+      HttpMediaTypeNotSupportedException.class,
+      HttpMediaTypeNotAcceptableException.class,
+      MissingPathVariableException.class,
+      MissingServletRequestParameterException.class,
+      MissingServletRequestPartException.class,
+      ServletRequestBindingException.class,
+      TypeMismatchException.class,
+      HttpMessageNotReadableException.class,
+      MethodArgumentNotValidException.class,
+      NoHandlerFoundException.class
+  })
```
왜 이거만 따로 빼나요? 쟤네를 잘 몰라서여

eggnee replied:
이 예외들은 INTERNAL SERVER ERROR 보다는 클라이언트 요청에 문제가 있을 때 발생하는 BAD_REQUEST 예외라서 따로 처리합니당

chs98412 commented on src/main/java/com/switcheese/server/common/exception/ExceptionMapper.java:
```diff
@@ -0,0 +1,49 @@
+import org.springframework.web.multipart.support.MissingServletRequestPartException;
+import org.springframework.web.servlet.NoHandlerFoundException;
+
+// TODO enum으로 변경
+public class ExceptionMapper {
+
+  public static HttpStatus toHttpStatus(Exception exception) {
```
어떻게 개선되나요 추후에

eggnee replied:
지금은 코드가 매우.. ^^ 가독성이 떨어지는데 enum으로 하면 가독성이 좋아집니당!
각자 맡은 부분 구현시에는 신경 안 쓰셔도 돼요

ghyen commented on src/main/java/com/switcheese/server/common/exception/BusinessException.java:
```diff
@@ -0,0 +1,33 @@
+@Getter
+public abstract class BusinessException extends RuntimeException {
+
+  protected HttpStatus status;
```
status도 불변하게 쓰일 것 같은데 code, message와 다르게 final을 안한 이유가 있나요?

ghyen replied:
코드 작성 깔끔하게 잘 하신 것 같아요! 나머지 수정하고 바로 머지해도 될 것 같아요!
